### DISCUSSION TITLE
Not receiving subscribed-to signals

### DISCUSSION THREAD

**Discussion by cnbaker** [Kistler-Group/sdbus-cpp#373] [Q&A] [answered]
Hi there. I'm attempting to subscribe to messages from the NetworkManager on the system bus when the wireless properties change. [Documentation is here](https://developer-old.gnome.org/NetworkManager/stable/gdbus-org.freedesktop.NetworkManager.Device.Wireless.html).

While searching these discussions, I found #161 which hinted to me to create a connection and enterEventLoop() at the end. By all indications, the following code _should_ work but I am simply not receiving the signal, sudo or not. I have built and using the latest v1.4.0 running on Debian 12 aarch64 GNU/Linux on a Raspberry Pi 4.

```c++
#include <sdbus-c++/sdbus-c++.h>
#include <vector>
#include <map>
#include <iostream>

const std::string dest = "org.freedesktop.NetworkManager";
const std::string iface = "org.freedesktop.NetworkManager";
const std::string d_iface = iface + ".Device";

int main(int argc, char** argv) {
	auto connection = sdbus::createSystemBusConnection();
	auto nm_proxy = sdbus::createProxy(*connection, dest, iface);
	std::vector<sdbus::ObjectPath> paths = nm_proxy->getProperty("Devices").onInterface(iface);
	std::string device_path;
	for (auto dp = paths.begin(); dp != paths.end(); ++dp) {

		auto d_proxy = sdbus::createProxy(*connection, dest, *dp);
		uint32_t dtype = d_proxy->getProperty("DeviceType").onInterface(d_iface);

		if (dtype == 2) {
			std::cout <<"found wireless device: " << *dp << std::endl;
			device_path = *dp;
			break;
		}
	}
	if (device_path.empty()) {
		std::cerr << "no wireless device found." << std::endl;
		return 1;
	}

	auto d_proxy = sdbus::createProxy(*connection, dest, device_path);
	d_proxy->uponSignal("PropertiesChanged").onInterface("org.freedesktop.DBus.Properties")
		.call([](std::string key, std::map<std::string, sdbus::Variant> change, std::map<std::string, sdbus::Variant> invalidated) {
				std::cout << "Wireless PropertiesChanged" << std::endl;
			});
	d_proxy->finishRegistration();
	connection->enterEventLoop();
}
```

While that code is running, I can see what I think is the subscribe request and signals to which I've subscribed, but am not receiving, using dbus-monitor --system:

```
...
method call time=1698624234.037048 sender=:1.790 -> destination=org.freedesktop.DBus serial=6 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=AddMatch
   string "type='signal',sender='org.freedesktop.NetworkManager',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged',path='/org/freedesktop/NetworkManager/Devices/3'"
method return time=1698624234.037291 sender=org.freedesktop.DBus -> destination=:1.790 serial=3 reply_serial=6
...
signal time=1698615612.468655 sender=:1.8 -> destination=(null destination) serial=1147 path=/org/freedesktop/NetworkManager/Devices/3; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.freedesktop.NetworkManager.Device.Wireless"
   array [
      dict entry(
         string "Bitrate"
         variant             uint32 24000
      )
   ]
   array [
   ]
```

I feel like I have distilled the problem to the simplest possible case, yet cannot seem to figure out what I'm doing wrong here. I'm sure it's quite simple and any guidance anyone can provide is greatly appreciated. Thank you!

**Comment by cnbaker** [comment]
Sorry, I have no idea why the code is not formatting correctly.

**Comment by sangelovic** [comment, marked as answer]
Looking at the code (markdown description of which I modified to fix the formatting problem), I see two problems:

1. `nm_proxy` is created with wrong third argument. It shall be an object path, not an interface. This by itself shall have thrown `sdbus::Error` exception already at this point.
2. The `invalidated` parameter of the "PropertiesChanged" is of type `as`, i.e. `std::vector<std::string>`, not `std::map<std::string, sdbus::Variant>`. Wrong argument type will result in `sdbus::Error` exception being thrown in the event loop thread, which in your case is also your main thread. So you should definitely see that exception.

Let me know if this helped.

**Reply by cnbaker** [reply]
Hi sangelovic. Thanks for your reply. I edited the posted code after pasting it, which is why createProxy had the wrong third argument 😳

After changing the callback signature to std::vector<std::string> for invalidated, I am now seeing the 'PropertiesChanged' in the callback which is great! Thank you for the hint. However, I haven't seen any exceptions raised when it is std::map<std::string, sdbus::Variant>. If I had seen those, I likely would have been able to deduce the issue, knowing at least the callback was called. Could it be that I haven't built the library properly or could there be some other issue?

Thank you!

**Reply by sangelovic** [reply]
Hi, no, no other issue or library built incorrectly.

I was actually incorrect regarding the exception in the event loop thread. I forgot the fact that these exceptions are caught by sdbus-c++ and ignored in order to not be propagated down to the C world. And I also forgot that in the past I already came up with one solution for this frequent issue with signal handler signature mismatch. It's an additional error argument, and it is described in the tip at the end of [this section](https://github.com/Kistler-Group/sdbus-cpp/blob/master/docs/using-sdbus-c%2B%2B.md#client-side-1) in the tutorial. This argument would reveal to you that the callback handler was called but a deserialization error occurred.

This discussion actually prompted me to improve a bit and unify handling of (all types of) exceptions from callback handlers, and fix one little bug, resulting in PR #375 that I've just prepared. As a side effect, exceptions from signal processing (including the deserialization) are caught and forwarded as `sd_bus_error` to the underlying sd-bus, which is much nicer than masking them.

sd-bus then [logs an error message](https://github.com/systemd/systemd/blob/main/src/libsystemd/sd-bus/bus-internal.c#L323) in such case. But we don't see this message because the default log severity is INFO, while this message is a DEBUG one.

The PR also mentions another way to approach problems with exceptions from message processing handlers -- logging this at sdbus-c++ level by calling a log callback that can be installed by clients via sdbus-c++ API.

**Reply by cnbaker** [reply]
Aha! This makes more sense now. I'll definitely try using those error args on callbacks going ahead. That log callback sounds like a fantastic idea. I'll keep an eye out for it. Thank you!
